### PR TITLE
Change post/preprocessing for taus.

### PR DESCRIPTION
**Pull request by Thammachath**
-Add transformation to Tau_idDeepTau2017v2p1VSjet, Tau_idDeepTau2017v2p1VSmu and Tau_idDeepTau2017v2p1VSe
-Add unsmear_decayMode
-Add postprocessing for decayMode

### CODE REVIEW

**Comment by francesco-vaselli** [comment]
Dear Thammachath,
Thanks for the pull request and sorry for the delay in our answer. Results and code are mostly fine, however we still have some doubts about your pre-processing of MTau_idDeepTau2017v2p1VSe and similar variables. 
Left a comment,
Best
Francesco & Filippo

File: trainer/extractor/taus/prep_actions.py [hunk 1]
```diff
@@ -40,11 +40,20 @@
         ["manual_range", [0.1, 5]],
         ["t", np.arctan, [10, -10]],
     ],
-    "MTau_idDeepTau2017v2p1VSe": [["u", 0.5, None]],
-    "MTau_idDeepTau2017v2p1VSjet": [["u", 0.5, None]],
+    "MTau_idDeepTau2017v2p1VSe": [
+        ["t", np.log2, [1, 1]],
```
Comment (francesco-vaselli) [inline trainer/extractor/taus/prep_actions.py:44]: Is np.log2 a way to get the values of the distribution closer together?
As this is not a linear transform, the uniform smearing afterwards is not acting as we would like it to.
Perhaps a better approach would be to *manually* move the values to specific ones so e.g. move 4 to 1, 8 to 2, 32 to 3 etc... And then apply the smearing.
Let us know if this is clear!

**Comment by Thammachath** [comment]
Dear Francesco & Filippo,

About the pre-processing of MTau_idDeepTau2017v2p1VSe and similar variables
In NanoAOD v9 bitmask of idDeepTau2017v2p1VSe are 1, 2, 4, 8, 16, 32, 64, 128 and 256 for Tau that pass VVTight.
which you can see from 
https://cms-nanoaod-integration.web.cern.ch/autoDoc/NanoAODv9/2018UL/doc_TTToSemiLeptonic_TuneCP5_13TeV-powheg-pythia8_RunIISummer20UL18NanoAODv9-106X_upgrade2018_realistic_v16_L1v1-v1.html#Tau

but the data from NanoAOD v9 root file bitmask of idDeepTau2017v2p1VSe are 0, 1, 3, 7, 15, 31, 63, 121 and 255. 
So, I decided +1 to all bitmask to match the reference. 

For NanoAOD v11 bitmask of idDeepTau2017v2p1VSe are 1, 2, 3, 4, 5, 6, 7, 8 and 9.
So, Phat and I think that convert to 1, 2, 3... (use binary logarithm) will be a good move for the future support.
 
However, the results after convert are 0, 1, 2, 3, 4, 5, 6, 7 and 8. 
I wanted to seek your guidance for the value conversion. Can I use a format of 1, 2, 3, 4... or should I stick to the original format 1, 2, 4, 8... ?

Thank you for your guidance and pull request reviewing. 

Best
Thammachath

**Comment by francesco-vaselli** [comment]
Dear Thammachath,
You are right. We didn't know about the data being 0, 1, 3, 7, 15, 31 etc.
If that is the case, then log2(x+1) seems a good preprocessing operation. You can certainly use it for letting the network learn the distribution.

Regarding postprocessing (what to write in the final NanoAOD) we think it's best to stick to the v9 standard for now. So from the output of your network (which will be 0, 1, 2, 3, 4, 5, 6, 7 and 8) you can write the inverse function to get back the v9 values.

Let us know if you have other questions, and if you do not I think we can safely proceed with the merge.
Best
Francesco & Filippo

**Comment by Thammachath** [comment]
Dear Francesco & Filippo,

Thank you for your feedback and guidance on the preprocessing and postprocessing. I agree with the approach you've laid out regarding maintaining the standard for NanoAOD v9.

I will make the necessary adjustments to the post-processing to ensure it matches the original. Once done, I'll send another pull request for review.

Please let me know if there are any other considerations or changes required.

Best regards,

Thammachath

**Review by francesco-vaselli — APPROVED** [review]
I see that you are trying to keep things consistent by using the inverse_transform utility we already developed. I appreciate the effort.

We need to take into account the possibility that some of the outputs of the network could fall outside of the [0, 8] range. At the moment we saturate on FullSim ranges after postprocessing, so I don't think this is a problem, but maybe we could also think about implementing a custom function in the future which ensures that values are in [0,8] before applying np.power. Just wanted to point this out.

Good job!

**Pull request merged**